### PR TITLE
rpcutil(engine): refactor rpc hook to interceptor

### PR DESCRIPTION
**Pull request by gozssky**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflow/issues/7695

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: engine/servermaster/server.go [hunk 1]
```diff
@@ -525,32 +453,94 @@ func (s *Server) initElector() error {
...
 	})
 
-	grpcServer := s.createGRPCServer()
-	defer grpcServer.Stop()
 	errGroup.Go(func() error {
-		return grpcServer.Serve(tcpServer.GrpcListener())
+		<-ctx.Done()
+		cleanup()
+		return errors.Trace(ctx.Err())
 	})
```
Comment (amyangfei) [inline engine/servermaster/server.go:542]: Is it necessary to check `ctx.Done()` here, other error group must have checked it. If this block can be removed, `cleanup` doesn't need to be protected with `cleanupOnce`

File: engine/servermaster/server.go [hunk 2]
```diff
@@ -68,31 +71,19 @@ import (
 const (
 	leaderElectionTable  = "leader_election"
 	resignLeaderDuration = 10 * time.Second
+	grpcConnBufSize      = 32 * 1024
 )
 
-// use a slice instead of map because in small data size, slice search is faster
-// than map search.
-var masterRPCLimiterAllowList = []string{
```
Comment (amyangfei) [inline engine/servermaster/server.go:75]: This log filter seems missing, do we need to add a log middleware

Comment (gozssky) [reply engine/servermaster/server.go:75]: We have log verbose request/response in error middleware.

File: engine/servermaster/server.go [hunk 3]
```diff
@@ -525,32 +453,94 @@ func (s *Server) initElector() error {
...
+			log.Warn("failed to close bufconn", zap.Error(err))
+		}
+	})
+	inProcessGRPCServer := s.createGRPCServer()
+	cleaupFuncs = append(cleaupFuncs, inProcessGRPCServer.Stop)
+	errGroup.Go(func() error {
+		return inProcessGRPCServer.Serve(bufConn)
+	})
+
+	dial := func(context.Context, string) (net.Conn, error) { return bufConn.Dial() }
```
[the remark this answers is not in the document]
Comment (gozssky) [reply engine/servermaster/server.go:509]: `bufConn` doesn't have `DialContext` method.

File: engine/servermaster/server.go (hunk 1 above)
Comment (gozssky) [reply engine/servermaster/server.go:544]: Some background goroutines may still be running after context is canceled. We need to explicitly stop or close them and wait for them to exit.

File: engine/servermaster/server.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (amyangfei) [reply engine/servermaster/server.go:524]: It has one

https://github.com/grpc/grpc-go/blob/09fc1a349826f04420292e0915fd11a5ce3dc347/test/bufconn/bufconn.go#L96

File: engine/servermaster/server.go (hunk 2 above)
Comment (amyangfei) [reply engine/servermaster/server.go:75]: The error middleware records
- verbose request if error happens
- debug log for normal request/response
But it lacks info logs for some critical method (without error), such as `CreateJob`, `CancelJob` etc.

File: engine/servermaster/server.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gozssky) [reply engine/servermaster/server.go]: Addressed in https://github.com/pingcap/tiflow/pull/7703/commits/6c50d649c68f83d2ae12548d9bfb20aae55ccc78.

File: engine/servermaster/server.go (hunk 2 above)
Comment (gozssky) [reply engine/servermaster/server.go:75]: Oh, I missed it. I will add it back.

Comment (gozssky) [reply engine/servermaster/server.go]: Added in https://github.com/pingcap/tiflow/pull/7703/commits/4830331e5d84e5bd847cb744148824f06c534190.

**Review by amyangfei — APPROVED** [review]

**Comment by amyangfei** [comment]
Unit test failed

```go
[2022-11-28T11:00:43.272Z] === Failed
[2022-11-28T11:00:43.272Z] === FAIL: engine/pkg/rpcutil TestCheckAvailable (0.00s)
[2022-11-28T11:00:43.272Z]     middleware_test.go:203: 
[2022-11-28T11:00:43.272Z]         	Error Trace:	/home/jenkins/agent/workspace/ut-check/tiflow/engine/pkg/rpcutil/middleware_test.go:203
[2022-11-28T11:00:43.272Z]         	Error:      	Should be true
[2022-11-28T11:00:43.272Z]         	Test:       	TestCheckAvailable
[2022-11-28T11:00:43.272Z] 
[2022-11-28T11:00:43.272Z] DONE 433 tests, 5 skipped, 1 failure in 330.148s
```

File: engine/servermaster/service_util.go [hunk 4]
```diff
@@ -16,47 +16,141 @@ package servermaster
...
+		return f.leaderCli, nil
+	}
+	f.rwm.RUnlock()
+
+	f.rwm.Lock()
+	defer f.rwm.Unlock()
+
+	if f.conn != nil {
+		if f.conn.Target() == leaderAddr {
+			return f.leaderCli, nil
```
Comment (CharlesCheung96) [inline engine/servermaster/service_util.go:126]: Is the double check here unnecessary?

**Review by maxshuang — APPROVED** [review]

**Pull request merged**